### ISSUE TITLE
[docs] Document how to connect to specific pod using connect

### ISSUE THREAD

**Issue by LaumiH** [hashicorp/consul-k8s#373]
**Edit from @lkysow:** See https://github.com/hashicorp/consul-k8s/issues/373#issuecomment-769351345 for how to accomplish this. This issue now tracks documenting the process properly.

---

If I understand correctly, Consul for now can only resolve a service of type StatefulSet via `<service>.service.<datacenter>.consul`.

What I am missing is the possibility to address the individual pods of the set via their Headless Service, as it is possible with K8s DNS <pod-0>.<service>.<namespace>.svc.cluster.local.

My use case is a cross-cluster MongoDB deployment, where the individual members can be addressed via their Consul resolvable names across cluster boundaries.

Of course this is useful with any StatefulSet and not limited to databases.

Explaining further what I want to achieve:
- setup Consul Federation with Mesh Gateways
- setup StatefulSet (e.g. MongoDB) in both clusters, annotate so Consul manages service
- annotate accordingly, so that upstream dependencies of form mongo-0.mongo.service.dc1.consul can be found connecting to various ports in each pod, e.g. `'consul.hashicorp.com/connect-service-upstreams': 'mongo-0:mongo:1234:dc1,mongo-1:mongo:5678:dc2'`
- use Consul to connect the mongo pods across clusters

Additionally, a sidecar would be helpful to configure all this automatically and set the replicaset inside mongo, but this is a different topic :)

I am also referring to [this forum post](https://discuss.hashicorp.com/t/kubernetes-service-mesh-access-to-service-instances/15218/6), where a user asked for the same feature.

**Comment by lkysow** [comment]
Hi Laura,
This might be possible if we can use the filter expression of the service resolver (https://www.consul.io/docs/agent/config-entries/service-resolver#filter) to select a service by its ID. The ID of the service in consul is `<service-name>-<pod id>`, e.g. `consul-consul-server-0`.

Perhaps you can test this out?

**Comment by LaumiH** [comment]
Thanks for the quick response! I will look into it now.

**Comment by LaumiH** [comment]
Ok, so I'm stuck and need help. I am new to Consul and cannot transfer what I am reading about the service resolver into k8s usage.

My understanding is as follows: 
I have created a Headless Service for my mongo StatefulSet, which gets [synced](https://www.consul.io/docs/k8s/service-sync#clusterip) into Consul by default or if I annotate it accordingly. 
In K8s I would now access the unique pods via `mongo-0.mongo.<namespace>.svc.cluster.local`. Of course this does not help in across-cluster StatefulSets.
I set the [service upstreams](https://www.consul.io/docs/k8s/connect#consul-hashicorp-com-connect-service-upstreams) of the mongo StatefulSet to `mongo:<port0>:dc1,mongo:<port1>:dc2`, but then I only get the mongo pods randomly. Here I would need to access the individual pods explicitly and deterministicly.

My question: how do I work with the service resolver filter in K8s? Can you please elaborate? :)
Should I create a Consul Service instead and let it sync with K8s, so the other way round?

**Comment by LaumiH** [comment]
But if Consul does not support StatefulSets and Headless Services, how do the Consul servers themselves connect to each other across cluster boundaries?

**Comment by mcortinas** [comment]
Hi,
 I'm working with GKE (k8s in GP) and headless services more than 3 years and I totally agree with Laura about its a big pity consul-k8s has no the ability of register specific POD name in statefulset

 Let me answer to Laura that Consul such us other technologies like Cassandra, Kafka, Elasticsearch has the ability the maintain the Cluster Metada by himself and just need a "discovery" enpoint in order to send to the clients(or other nodes in the backend) the entire TOPOLOGY of the cluster(all nodesID and IPS), this is a logic implemented by these technologies.
  
 I remember I was using a python library 3 years ago due to Consul can not help me and Neo4j needed it. let me share...
 [kubernetes client python library](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/CoreV1Api.md#list_endpoints_for_all_namespaces)

 For other hand, let me add that it's a big pity that my favorite service discovery technoogy has no the feature of register POD name in statefulset, I think hashicorp should fix this.

 Best,
 Marc

**Comment by ishustava** [comment]
Hey @LaumiH and @mcortinas,

I tried out @lkysow suggestion above to use filtering and realized that consul-k8s will not work as-is. That is because the way we're creating the meta key for the pod name `pod-name` will not work with the Consul's filtering API, which does not support dashes. I've created  #427 to fix that.

I was able to use that fix and accomplish essentially what you are asking for. 

Here is what I did:

Installed Helm chart with (note the dev build of consul-k8s image I'm using and that I'm installing the helm chart from the current `master`):

```
helm install iryna --set global.imageK8S=ishustava/consul-k8s-dev:01-28-2021-402498b --set server.replicas=1 --set connectInject.enabled=true --set controller.enabled=true .
```

Applied proxy-defaults to make all services HTTP (otherwise, we cannot use L7 features):

```
apiVersion: consul.hashicorp.com/v1alpha1
kind: ProxyDefaults
metadata:
  name: global
spec:
  config:
    protocol: http
```

Created my stateful set for testing (it outputs "hello world, $POD_IP" so that I can see which instance I'm talking to):

```
apiVersion: v1
kind: Service
metadata:
  name: web
  labels:
    app: web
spec:
  ports:
    - port: 80
      name: web
  clusterIP: None
  selector:
    app: web
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  serviceName: "web"
  replicas: 2
  selector:
    matchLabels:
      app: web
  template:
    metadata:
      labels:
        app: web
      annotations:
        "consul.hashicorp.com/connect-inject": "true"
        "consul.hashicorp.com/connect-service": "web"
    spec:
      containers:
        - name: web
          env:
            - name: POD_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
          image: ishustava/http-echo:latest
          command:
              - "/bin/sh"
              - "-ec"
              - |
                /http-echo -listen=:80 -text="hello world $POD_IP"
          ports:
            - containerPort: 80
              name: http
          volumeMounts:
            - name: www
              mountPath: /usr/share/web/html
  volumeClaimTemplates:
    - metadata:
        name: www
      spec:
        accessModes: [ "ReadWriteOnce" ]
        resources:
          requests:
            storage: 1Gi
```

Create service resolver so that my statefulset service can have a subset for each instance. I used `pod_name` metadata key for that:

```
apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceResolver
metadata:
  name: web
spec:
  subsets:
    'web0':
      filter: "Service.Meta.pod_name == \"web-0\""
    'web1':
      filter: "Service.Meta.pod_name == \"web-1\""
```

Create two virtual services called `web-0` and `web-1` and point them to the corresponding subset of the `web` service:

```
apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceResolver
metadata:
  name: web-0
spec:
  redirect:
    service: web
    serviceSubset: web0
---
apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceResolver
metadata:
  name: web-1
spec:
  redirect:
    service: web
    serviceSubset: web1
```

Create my client service where I will use these virtual services as upstream (notice my connect upstream annotations):
```
apiVersion: v1
kind: ServiceAccount
metadata:
  name: static-client
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: static-client
spec:
  replicas: 1
  selector:
    matchLabels:
      app: static-client
  template:
    metadata:
      name: static-client
      labels:
        app: static-client
      annotations:
        "consul.hashicorp.com/connect-inject": "true"
        "consul.hashicorp.com/connect-service-upstreams": "web-0:1234,web-1:1235"
    spec:
      containers:
        - name: static-client
          image: tutum/curl:latest
          command: [ "/bin/sh", "-c", "--" ]
          args: [ "while true; do sleep 30; done;" ]
      serviceAccountName: static-client
```

Note the two IPs of my `web` service:
```
$ kubectl get pods -o wide
NAME                                                              READY   STATUS    RESTARTS   AGE    IP           NODE                                      NOMINATED NODE   READINESS GATES
iryna-consul-ccssm                                                1/1     Running   0          16m    10.24.1.92   gke-shustava-default-pool-bad0f39a-08fj   <none>           <none>
iryna-consul-connect-injector-webhook-deployment-f56fcc696pg8xm   1/1     Running   0          16m    10.24.3.28   gke-shustava-default-pool-bad0f39a-9jwd   <none>           <none>
iryna-consul-controller-79cfb9795c-rtwmn                          1/1     Running   0          16m    10.24.3.30   gke-shustava-default-pool-bad0f39a-9jwd   <none>           <none>
iryna-consul-fjjcz                                                1/1     Running   0          16m    10.24.2.48   gke-shustava-default-pool-bad0f39a-f890   <none>           <none>
iryna-consul-gpzwt                                                1/1     Running   0          16m    10.24.4.30   gke-shustava-default-pool-bad0f39a-t1z4   <none>           <none>
iryna-consul-mhr7q                                                1/1     Running   0          16m    10.24.3.29   gke-shustava-default-pool-bad0f39a-9jwd   <none>           <none>
iryna-consul-server-0                                             1/1     Running   0          16m    10.24.4.32   gke-shustava-default-pool-bad0f39a-t1z4   <none>           <none>
iryna-consul-webhook-cert-manager-56b9bc5bd7-lvldp                1/1     Running   0          16m    10.24.4.31   gke-shustava-default-pool-bad0f39a-t1z4   <none>           <none>
iryna-consul-zcx6c                                                1/1     Running   0          16m    10.24.0.57   gke-shustava-default-pool-bad0f39a-wr0v   <none>           <none>
static-client-65946b487f-6d26v                                    3/3     Running   0          32s    10.24.3.33   gke-shustava-default-pool-bad0f39a-9jwd   <none>           <none>
web-0                                                             3/3     Running   0          100s   10.24.1.94   gke-shustava-default-pool-bad0f39a-08fj   <none>           <none>
web-1                                                             3/3     Running   0          77s    10.24.4.34   gke-shustava-default-pool-bad0f39a-t1z4   <none>           <none>
```

I can now exec into my static-client pod and try to talk to the two upstreams: 
```
$ kubectl exec -it static-client-65946b487f-6d26v -- /bin/sh
Defaulting container name to static-client.
Use 'kubectl describe pod/static-client-65946b487f-6d26v -n default' to see all of the containers in this pod.
# curl localhost:1234
hello world 10.24.1.94
# curl localhost:1235
hello world 10.24.4.34
```

Let me know if this helps with your use cases!

**Comment by LaumiH** [comment]
This essentially does exactly what I was looking for, thank you so much @ishustava !
My follow-up question would be if this can be included as an "accepted workaround" somewhere in the docs. I am almost sure that this will help other users as well. Addressing individual Pods might not be considered cloud native really, but there are more and more efforts to deploy databases in k8s distributed across geographies, and essentially being able to address the Pods is the only complex thing that is challenging to implement.

**Comment by lkysow** [comment]
Agreed we should definitely document this. I'm going to update this issue to track the documentation.

**Comment by josvegit** [comment]
This doesn't seem to work for other stuff the **http** though. For example If I run a mongo-replicaset instead of that http-echoserver and try to connect to one of those mongo servers using mongo cli (with the same upsteams as static-client), then it doesn't work.

**Comment by josvegit** [comment]
I was able to run above example from @ishustava but when I change from an http-echoserver to a repliaset with 2 mongos. Then change. the static-client to a container supporting the mongo-cli then I cant connect to those pods(mongo-1 and mongo-2). Im guessing this is because mongo uses a lower level protocol ?

**Comment by david-yu** [comment]
Hi @josvegit and @LaumiH we now have direct support for this use case as of Consul 1.10 via Transparent Proxy. We just released a new Helm chart 0.32 and Consul 0.26 k8s binary to support a new field within ServiceDefaults CRD to dial individual instances of a service directly. See [Headless Services](https://www.consul.io/docs/connect/transparent-proxy#headless-services) within our TProxy docs!

**Comment by nflaig** [comment]
Hi all, anyone is running mongodb as replica set with transparent proxy enabled? I am facing issues with the sync of the replicas they are not able to connect to each other. @david-yu is there an example on how to run stateful sets with transparent proxy enabled? I also created a [discussion in the forum](https://discuss.hashicorp.com/t/transparent-proxy-with-stateful-sets/30169) related to using tproxy with stateful sets.
